nexus/src/db/datastore/project.rs

```diff
@@ -65,33 +66,241 @@ impl DataStore {
+                .limit(1)
+                .first_async::<Uuid>(self.pool_authorized(opctx).await?)
+                .await,
+        )
+        .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))?;
+        if ip_pool_found.is_some() {
+            return Err(Error::InvalidRequest {
+                message: "project to be deleted contains an ip pool"
+                    .to_string(),
+            });
+        }
+
+        Ok(())
+    }
+
+    async fn ensure_no_external_ips_in_project(
+        &self,
+        opctx: &OpContext,
+        authz_project: &authz::Project,
+    ) -> DeleteResult {
+        use db::schema::external_ip;
+
+        // Make sure there are no external_ips present within this project.
+        let external_ip_found = diesel_pool_result_optional(
+            external_ip::dsl::external_ip
+                .filter(external_ip::dsl::project_id.eq(authz_project.id()))
+                .filter(external_ip::dsl::time_deleted.is_null())
+                .select(external_ip::dsl::id)
+                .limit(1)
+                .first_async::<Uuid>(self.pool_authorized(opctx).await?)
+                .await,
+        )
+        .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))?;
+        if external_ip_found.is_some() {
+            return Err(Error::InvalidRequest {
+                message: "project to be deleted contains an external IP"
+                    .to_string(),
+            });
+        }
+
+        Ok(())
+    }
+
     /// Delete a project
-    // TODO-correctness This needs to check whether there are any resources that
-    // depend on the Project (Disks, Instances).  We can do this with a
-    // generation counter that gets bumped when these resources are created.
     pub async fn project_delete(
         &self,
         opctx: &OpContext,
         authz_project: &authz::Project,
+        db_project: &db::model::Project,
     ) -> DeleteResult {
         opctx.authorize(authz::Action::Delete, authz_project).await?;
 
+        // Verify that child resources do not exist.
+        self.ensure_no_instances_in_project(opctx, authz_project).await?;
+        self.ensure_no_disks_in_project(opctx, authz_project).await?;
+        self.ensure_no_images_in_project(opctx, authz_project).await?;
+        self.ensure_no_snapshots_in_project(opctx, authz_project).await?;
+        self.ensure_no_vpcs_in_project(opctx, authz_project).await?;
+        self.ensure_no_ip_pools_in_project(opctx, authz_project).await?;
+        self.ensure_no_external_ips_in_project(opctx, authz_project).await?;
+
         use db::schema::project::dsl;
 
         let now = Utc::now();
-        diesel::update(dsl::project)
+        let updated_rows = diesel::update(dsl::project)
             .filter(dsl::time_deleted.is_null())
             .filter(dsl::id.eq(authz_project.id()))
+            .filter(dsl::rcgen.eq(db_project.rcgen))
             .set(dsl::time_deleted.eq(now))
```

Comment:
Does this not also update the rcgen?

Reply:
Nah, it doesn't need to. The whole point of the rcegen number is that it allows us to do:

1. Check if children exist
2. Try to update project as "time_deleted.eq(now)"

As if those two operations were **atomic**.

- If a new child *was* inserted when we do (2) the rcgen should change. This is the only spot we "read" the rcgen.
- If no new children are inserted when we do (2), the operation can know that (because it already checked) no children previously existed, and no children exist now.

So we don't need to update the value in this code path -- if we filter this record and see any project, it's gonna be deleted anyway!